chain/chain/src/chain.rs

```diff
@@ -3067,14 +3087,24 @@ impl<'a> ChainUpdate<'a> {
         }
 
         // Verify that proposals from chunks match block header proposals.
-        let mut all_chunk_proposals = vec![];
-        for chunk in block.chunks().iter() {
-            if block.header().height() == chunk.height_included() {
-                all_chunk_proposals.extend(chunk.validator_proposals().iter().cloned());
+        let block_height = block.header().height();
+        for (hp, cp) in utils::full_zip(
+            block.header().validator_proposals(),
+            block
+                .chunks()
+                .iter()
+                .filter(|chunk| block_height == chunk.height_included())
+                .flat_map(|chunk| chunk.validator_proposals()),
+        ) {
+            // Can only occur if there were a different number of proposals in the header
+            // and chunks
+            if hp.is_none() || cp.is_none() {
+                return Err(ErrorKind::InvalidValidatorProposals.into());
+            }
+            // Proposals differed!
+            if hp.unwrap() != cp.unwrap() {
+                return Err(ErrorKind::InvalidValidatorProposals.into());
```

Comment:
It's probably better to write this as a match to avoid using `unwrap`

Reply:
Done in https://github.com/near/nearcore/pull/4115/commits/95d8768defc681c1a6acf7d36615ce5b13081cfe